datafusion/datasource-parquet/src/row_group_filter.rs

```diff
@@ -135,13 +186,56 @@ impl RowGroupAccessPlanFilter {
         // try to prune the row groups in a single call
         match predicate.prune(&pruning_stats) {
             Ok(values) => {
-                // values[i] is false means the predicate could not be true for row group i
+                let mut fully_contained_candidates_original_idx: Vec<usize> = Vec::new();
                 for (idx, &value) in row_group_indexes.iter().zip(values.iter()) {
                     if !value {
                         self.access_plan.skip(*idx);
                         metrics.row_groups_pruned_statistics.add_pruned(1);
                     } else {
                         metrics.row_groups_pruned_statistics.add_matched(1);
+                        fully_contained_candidates_original_idx.push(*idx);
+                    }
+                }
+
+                // Note: this part of code shouldn't be expensive with a limited number of row groups
+                // If we do find it's expensive, we can consider optimizing it further.
+                if !fully_contained_candidates_original_idx.is_empty() {
+                    // Use NotExpr to create the inverted predicate
+                    let inverted_expr =
+                        Arc::new(NotExpr::new(Arc::clone(predicate.orig_expr())));
+                    // Simplify the NOT expression (e.g., NOT(c1 = 0) -> c1 != 0)
+                    // before building the pruning predicate
+                    let mut simplifier = PhysicalExprSimplifier::new(arrow_schema);
+                    let inverted_expr = simplifier.simplify(inverted_expr).unwrap();
+                    if let Ok(inverted_predicate) = PruningPredicate::try_new(
+                        inverted_expr,
+                        Arc::clone(predicate.schema()),
+                    ) {
+                        let inverted_pruning_stats = RowGroupPruningStatistics {
+                            parquet_schema,
+                            row_group_metadatas: fully_contained_candidates_original_idx
+                                .iter()
+                                .map(|&i| &groups[i])
+                                .collect::<Vec<_>>(),
+                            arrow_schema,
+                        };
+
+                        if let Ok(inverted_values) =
+                            inverted_predicate.prune(&inverted_pruning_stats)
+                        {
+                            for (i, &original_row_group_idx) in
+                                fully_contained_candidates_original_idx.iter().enumerate()
+                            {
+                                // If the inverted predicate *also* prunes this row group (meaning inverted_values[i] is false),
```

Comment:
This is a very clever idea -- basically if `not(predicate)` would prune the predicate it means that `predicate` evaluated to (known) `true` for all rows

I tried thinking of counter examples, but I couldn't come up with any and the reasoning is sound